travis.sh

```diff
@@ -23,8 +23,44 @@ SKIP_TESTS=false
 if [ -z "$GOOGLE_APPLICATION_CREDENTIALS" ] ; then
   SKIP_TESTS=true
 fi
+
+# Finds the closest parent dir that encompasses all changed files, and has a
+# pom.xml
+travis_changed_files_parent() {
+  [ -z "$TRAVIS_PULL_REQUEST" ] && return 0  # If we're not in a PR, forget it
+
+  (
+    set +e
+
+    changed="$(git diff --name-only "$TRAVIS_COMMIT" "$TRAVIS_BRANCH")"
+    if [ $? -ne 0 ]; then
+      # Fall back to git head
+      changed="$(git diff --name-only "$(git rev-parse HEAD)" "$TRAVIS_BRANCH")"
+      [ $? -ne 0 ] && return 0  # Give up. Just run everything.
+    fi
+
+    # Find the common prefix
+    prefix="$(echo "$changed" | grep -v travis.sh | sed -e 'N;s/^\(.*\).*\n\1.*$/\1\n\1/;D')"
```

Comment:
Magic! I might want a bit more explanation in the comments for the sed command. (Something like https://robots.thoughtbot.com/literate-vim but maybe not quite that detailed)